### PR TITLE
P2P bootstrap small refactor

### PR DESCRIPTION
**Pull request by iulianpascalau**
- modified the p2p.bootstrapping wait process to consider a minimum number of peers

### CODE REVIEW

File: testscommon/goroutines/filter.go [hunk 1]
```diff
@@ -9,6 +9,8 @@ var ignorable = []string{
 	"github.com/beevik/ntp.getTime",
 	"/usr/local/go/src/net/dnsclient_unix.go",
 	"/usr/local/go/src/runtime/proc.go",
+	"net._C2func_getaddrinfo",
```
Comment (iulianpascalau) [inline testscommon/goroutines/filter.go:12]: sometimes the tests that count go routines started "capture" some go routines that do other, unrelated stuff regarding the tests, such as DNS resolving

File: cmd/node/config/p2p.toml [hunk 2]
```diff
@@ -24,6 +24,10 @@
     #the sync and consensus mechanisms
     ThresholdMinConnectedPeers = 3
 
+    # WaitNumPeersOnBootstrap is the minimum number of peers to wait on bootstrap or the node will wait the default
+    # time which is now set to ~20 seconds
+    WaitNumPeersOnBootstrap = 10
```
Comment (mariusmihaic) [inline cmd/node/config/p2p.toml:29]: Perhaps rename it to  smth like:`MinNumPeersToWaitForOnBootstrap`?

Also, maybe it would be worth to add in the description what is the name of the variable which defines:
> the default time which is now set to ~20 seconds

since its value might be changed

File: factory/networkComponents.go [hunk 3]
```diff
@@ -168,7 +168,7 @@ func (ncf *networkComponentsFactory) Create() (*networkComponents, error) {
 		return nil, err
 	}
 
-	ncf.waitForBootstrap(ncf.bootstrapWaitSeconds)
+	netMessenger.WaitForConnections(time.Duration(ncf.bootstrapWaitSeconds)*time.Second, ncf.p2pConfig.Node.WaitNumPeersOnBootstrap)
```
Comment (mariusmihaic) [inline factory/networkComponents.go:171]: [optional]
If you find it suitable, perhaps extract this `time.Duration(ncf.bootstrapWaitSeconds)*time.Second` in a new variable (or declare a const time var/ or store it inside struct as time var field and initialize it in constructor)

File: p2p/libp2p/netMessenger.go [hunk 4]
```diff
@@ -686,6 +687,46 @@ func (netMes *networkMessenger) Bootstrap() error {
...
 
+// WaitForConnections will wait the maxWaitingTime duration or until the target connected peers was achieved
+func (netMes *networkMessenger) WaitForConnections(maxWaitingTime time.Duration, minNumOfPeers uint32) {
+	startTime := time.Now()
+	defer func() {
+		log.Debug("networkMessenger.WaitForConnections",
+			"waited", time.Since(startTime), "num connected peers", len(netMes.ConnectedPeers()))
+	}()
+
+	if minNumOfPeers <= 0 {
```
Comment (mariusmihaic) [inline p2p/libp2p/netMessenger.go:698]: Since `minNumOfPeers` is a `uint32` variable, checking if it is `<= 0` is redundant, since it cannot be negative. 
Perhaps change it to  `minNumOfPeers == 0` ?

File: p2p/libp2p/netMessenger.go [hunk 5]
```diff
@@ -686,6 +687,46 @@ func (netMes *networkMessenger) Bootstrap() error {
...
+		time.Sleep(maxWaitingTime)
+		log.Debug("networkMessenger.WaitForConnections", "waiting", maxWaitingTime)
```
Comment (mariusmihaic) [inline p2p/libp2p/netMessenger.go:700]: Perhaps change the order of these two lines?
Right now you `Sleep(maxWaitingTime)` and upon finishing waiting you add a message in logger that you are waiting, when in fact you actually finished waiting

File: p2p/libp2p/netMessenger_test.go [hunk 6]
```diff
@@ -1799,3 +1799,50 @@ func TestNetworkMessenger_Bootstrap(t *testing.T) {
...
+
+		_ = mes1.ConnectToPeer(mes2.Addresses()[0])
+		go func() {
+			time.Sleep(time.Second * 2)
+			_ = mes1.ConnectToPeer(mes3.Addresses()[0])
+		}()
+
+		defer func() {
+			_ = mes1.Close()
+			_ = mes2.Close()
```
Comment (mariusmihaic) [inline p2p/libp2p/netMessenger_test.go:1839]: Should we also add this: ?
`_ = mes3.Close()`

File: p2p/libp2p/netMessenger_test.go [hunk 7]
```diff
@@ -1799,3 +1799,50 @@ func TestNetworkMessenger_Bootstrap(t *testing.T) {
...
+		}()
+
+		timeToWait := time.Hour
```
Comment (mariusmihaic) [inline p2p/libp2p/netMessenger_test.go:1842]: Why one **hour** here?

File: p2p/libp2p/netMessenger_test.go [hunk 8]
```diff
@@ -1799,3 +1799,50 @@ func TestNetworkMessenger_Bootstrap(t *testing.T) {
 	goRoutinesNumberStart := runtime.NumGoroutine()
 	core.DumpGoRoutinesToLog(goRoutinesNumberStart, log)
 }
+
+func TestNetworkMessenger_WaitForConnections(t *testing.T) {
```
Comment (mariusmihaic) [inline p2p/libp2p/netMessenger_test.go:1803]: Shall we add another test in which no peer will connect within the waiting time-frame? Or are there other tests in this file that take care of it?

**Review by sasurobert — DISMISSED** [review]

File: p2p/libp2p/netMessenger_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go:1839]: done

File: cmd/node/config/p2p.toml (hunk 2 above)
Comment (iulianpascalau) [reply cmd/node/config/p2p.toml:29]: done

File: factory/networkComponents.go (hunk 3 above)
Comment (iulianpascalau) [reply factory/networkComponents.go:171]: done

File: p2p/libp2p/netMessenger.go (hunk 4 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger.go:698]: right, remained from the initial implementation in which I've used an int

File: p2p/libp2p/netMessenger.go (hunk 5 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger.go:700]: done

File: p2p/libp2p/netMessenger_test.go (hunk 6 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go:1839]: done

File: p2p/libp2p/netMessenger_test.go (hunk 7 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go:1842]: switched to 1minute which will not cause the package test to fail with timeout (15 minutes)

File: p2p/libp2p/netMessenger_test.go (hunk 8 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go:1803]: added test

File: cmd/node/config/p2p.toml [hunk 9]
```diff
@@ -24,9 +24,9 @@
     #the sync and consensus mechanisms
     ThresholdMinConnectedPeers = 3
 
-    # WaitNumPeersOnBootstrap is the minimum number of peers to wait on bootstrap or the node will wait the default
-    # time which is now set to ~20 seconds
-    WaitNumPeersOnBootstrap = 10
+    # MinNumPeersToWaitForOnBootstrap is the minimum number of peers to wait on bootstrap or the node will wait the default
+    # time which is now set to ~20 seconds (the const defined in the common package named TimeToWaitForP2PBootstrap)
```
Comment (mariusmihaic) [inline cmd/node/config/p2p.toml:28]: :+1:

File: p2p/libp2p/netMessenger_test.go [hunk 10]
```diff
@@ -1837,12 +1837,31 @@ func TestNetworkMessenger_WaitForConnections(t *testing.T) {
...
 		}()
 
-		timeToWait := time.Hour
+		timeToWait := time.Second * 10
 		mes1.WaitForConnections(timeToWait, 2)
 
 		assert.True(t, timeToWait > time.Since(startTime))
 		assert.True(t, libp2p.PollWaitForConnectionsInterval <= time.Since(startTime))
 	})
+	t.Run("min num of peers is 2 but we only connected to 1 peer", func(t *testing.T) {
```
Comment (mariusmihaic) [inline p2p/libp2p/netMessenger_test.go:1849]: :+1:

**Review by mariusmihaic — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

File: cmd/node/config/p2p.toml [hunk 11]
```diff
@@ -24,6 +24,12 @@
     #the sync and consensus mechanisms
     ThresholdMinConnectedPeers = 3
 
+    # ConnectionWatcherType represents the type of a connection watcher needed.
+    # possible options:
+    #  - "disabled" - no connection wathing should be made
```
Comment (mariusmihaic) [inline cmd/node/config/p2p.toml:29]: [comment]

I know this is not part of your PR, but perhaps rename the typo in `wathing` ?

**Review by mariusmihaic — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed .

**Review by sasurobert — APPROVED** [review]
